File: Python/graph_algorithms/08_minimum_cost_path.py
import sys
from math import inf, isinf
from collections import defaultdict


class DirectedGraph(object):
    def __init__(self, edge_list, nr_nodes):
        self._adjacency_list = self._convert_edgelist_to_adjacency_list(edge_list)
        self._nr_nodes = nr_nodes

    def _convert_edgelist_to_adjacency_list(self, edge_list):
        adjacency_list = defaultdict(list)
        for node_from, node_to, weight in edge_list:
            adjacency_list[node_from].append((node_to, weight))
        return dict(adjacency_list)

    def _update_distances(self, u, dist_u, distances, T):
        neighbours = self._adjacency_list.get(u, [])
        for v, weight in neighbours:
            distance_through_u = dist_u + weight
            if distances[v-1] > distance_through_u:
                distances[v-1] = distance_through_u
                T[v] = distance_through_u

    def dijkstra_algorithm(self, start_node):
        distances = [inf] * self._nr_nodes
        distances[start_node-1] = 0
        T = {node: distances[node-1] for node in range(1, self._nr_nodes+1)}
        while T:
            u = min(T, key=T.get)
            dist_u = T.pop(u)
            self._update_distances(u, dist_u, distances, T)
        return distances


def find_cheapest_path(edge_list, nr_nodes, s, t):
    graph = DirectedGraph(edge_list, nr_nodes)
    distances_from_s = graph.dijkstra_algorithm(s)
    result = distances_from_s[t-1]
    if isinf(result):
        return -1
    return result


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def read_weighted_edges(reader, nr_edges):
    edges = []
    for _ in range(nr_edges):
        edge = tuple(convert_to_intlist(next(reader)))
        edges.append(edge)
    return edges


def main():
    reader = sys.stdin
    nr_nodes, nr_edges = convert_to_intlist(next(reader))
    edge_list = read_weighted_edges(reader, nr_edges)
    s, t = convert_to_intlist(next(reader))
    result = find_cheapest_path(edge_list, nr_nodes, s, t)
    print(result)


if __name__ == "__main__":
    main()

File: Rust/graph_algorithms/src/bin/08_minimum_cost_path.rs
use std::cmp::{Ordering, Reverse};
use std::collections::{BinaryHeap, HashMap, HashSet};
use std::io;

#[derive(Clone)]
struct Edge {
    from: i32,
    to: i32,
    weight: i32,
}

struct DirectedGraph {
    nr_nodes: i32,
    adjacency_list: HashMap<i32, Vec<Edge>>
}

impl DirectedGraph {
    fn from_edge_list(nr_nodes: i32, edges: Vec<Edge>) -> Self {
        let mut adjacency_list = HashMap::new();
        for edge @ Edge { from, ..} in edges {
            adjacency_list.entry(from).or_insert_with(Vec::new).push(edge);
        }

        DirectedGraph { nr_nodes, adjacency_list }
    }
}

#[derive(Clone, Copy, Eq, PartialEq)]
enum Distance {
    Dist(i32),
    Infinity,
}

impl Distance {
    fn add(&self, other: &Self) -> Self {
        match (self, other) {
            (Distance::Dist(d1), Distance::Dist(d2)) => Distance::Dist(d1 + d2),
            _ => Distance::Infinity,
        }
    }
}

impl Ord for Distance {
    fn cmp(&self, other: &Self) -> Ordering {
        match (self, other) {
            (Distance::Infinity, Distance::Infinity) => Ordering::Equal,
            (Distance::Infinity, Distance::Dist(_)) => Ordering::Greater,
            (Distance::Dist(_), Distance::Infinity) => Ordering::Less,
            (Distance::Dist(d1), Distance::Dist(d2)) => d1.cmp(d2),
        }
    }
}

impl PartialOrd for Distance {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn convert_to_int_vector(line: &str) -> Vec<i32> {
    line
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect()
}

fn read_line() -> String {
    let mut line = String::new();
    io::stdin().read_line(&mut line).unwrap();

    line
}

fn read_pair() -> (i32, i32) {
    let line = read_line();
    let xs: Vec<i32> = convert_to_int_vector(&line);
    let [a, b] = xs.as_slice() else {
        panic!("Malformed input: {}", line);
    };
    (*a, *b)
}

fn read_edges(nr_edges: usize) -> Vec<Edge> {
    let mut edges = Vec::with_capacity(nr_edges);
    for _ in 0..nr_edges {
        let line = read_line();
        let xs: Vec<i32> = convert_to_int_vector(&line);
        let [a, b, w] = xs.as_slice() else {
            panic!("Malformed input: {}", line);
        };
        edges.push(Edge { from: *a, to: *b, weight: *w });
    }

    edges
}

fn find_cheapest_path(graph: DirectedGraph, start: i32, end: i32) -> Distance {
    let mut distances = vec![Distance::Infinity; graph.nr_nodes as usize];
    distances[start as usize - 1] = Distance::Dist(0);
    let mut heap = BinaryHeap::from([Reverse((Distance::Dist(0), start))]);
    let mut finalized_nodes = HashSet::new();
    loop {
        let Some(Reverse((dist, node))) = heap.pop() else {
            return distances[end as usize - 1]
        };
        if !finalized_nodes.contains(&node) {
            if let Some(neighbors) = graph.adjacency_list.get(&node) {
                for &Edge {to, weight, ..} in neighbors {
                    let distance_through_node = dist.add(&Distance::Dist(weight));
                    if distances[to as usize - 1] > distance_through_node {
                        distances[to as usize - 1] = distance_through_node;
                        heap.push(Reverse((distance_through_node, to)));
                    }
                }
                finalized_nodes.insert(node);
            };
        };
    };
}

fn main() {
    let (nr_nodes, nr_edges) = read_pair();
    let edge_list = read_edges(nr_edges as usize);
    let (s, t) = read_pair();
    let graph = DirectedGraph::from_edge_list(nr_nodes, edge_list);
    let result = find_cheapest_path(graph, s, t);
    match result {
        Distance::Infinity => println!("-1"),
        Distance::Dist(d) => println!("{}", d),
    };
}
